zcash_primitives/src/transaction/sighash_v5.rs

```diff
@@ -0,0 +1,179 @@
+use std::io::Write;
+
+use blake2b_simd::Hash as Blake2bHash;
+use byteorder::{LittleEndian, WriteBytesExt};
+
+use crate::transaction::{
+    blake2b_256::HashWriter,
+    components::transparent::{self, TxOut},
+    sighash::{
+        SignableInput, TransparentInput, SIGHASH_ANYONECANPAY, SIGHASH_MASK, SIGHASH_NONE,
+        SIGHASH_SINGLE,
+    },
+    txid::{
+        to_hash, transparent_outputs_hash, transparent_prevout_hash, transparent_sequence_hash,
+    },
+    Authorization, TransactionData, TransparentDigests, TxDigests,
+};
+
+#[cfg(feature = "zfuture")]
+use std::convert::TryInto;
+
+#[cfg(feature = "zfuture")]
+use crate::{
+    serialize::{CompactSize, Vector},
+    transaction::{components::tze, sighash::TzeInput, TzeDigests},
+};
+
+const ZCASH_TRANSPARENT_INPUT_HASH_PERSONALIZATION: &[u8; 16] = b"Zcash___TxInHash";
+
+#[cfg(feature = "zfuture")]
+const ZCASH_TZE_INPUT_HASH_PERSONALIZATION: &[u8; 16] = b"Zcash__TzeInHash";
+
+fn transparent_input_sigdigests<A: transparent::Authorization>(
+    bundle: &transparent::Bundle<A>,
+    input: &TransparentInput<'_>,
+    txid_digests: &TransparentDigests<Blake2bHash>,
+    hash_type: u32,
+) -> TransparentDigests<Blake2bHash> {
+    let flag_anyonecanpay = hash_type & SIGHASH_ANYONECANPAY != 0;
+    let flag_single = hash_type & SIGHASH_MASK == SIGHASH_SINGLE;
+    let flag_none = hash_type & SIGHASH_MASK == SIGHASH_NONE;
+
+    let prevout_digest = if flag_anyonecanpay {
+        transparent_prevout_hash::<A>(&[])
+    } else {
+        txid_digests.prevout_digest
+    };
+
+    let sequence_digest = if flag_anyonecanpay || flag_single || flag_none {
+        transparent_sequence_hash::<A>(&[])
+    } else {
+        txid_digests.sequence_digest
+    };
+
+    let outputs_digest = if flag_single {
+        if input.index() < bundle.vout.len() {
+            transparent_outputs_hash(&[&bundle.vout[input.index()]])
+        } else {
+            transparent_outputs_hash::<TxOut>(&[])
+        }
+    } else if flag_none {
+        transparent_outputs_hash::<TxOut>(&[])
+    } else {
+        txid_digests.outputs_digest
+    };
+
+    // If we are serializing an input (i.e. this is not a JoinSplit signature hash):
+    //   a. outpoint (32-byte hash + 4-byte little endian)
+    //   b. scriptCode of the input (serialized as scripts inside CTxOuts)
+    //   c. value of the output spent by this input (8-byte little endian)
+    //   d. nSequence of the input (4-byte little endian)
+    let mut ch = HashWriter::new(ZCASH_TRANSPARENT_INPUT_HASH_PERSONALIZATION);
+    let txin = &bundle.vin[input.index()];
+    txin.prevout.write(&mut ch).unwrap();
+    input.script_code().write(&mut ch).unwrap();
+    ch.write_all(&input.value().to_i64_le_bytes()).unwrap();
+    ch.write_u32::<LittleEndian>(txin.sequence).unwrap();
+    let per_input_digest = ch.finalize();
```

Comment:
Should there be an `if` condition around this block, checking if we're dealing with a transparent input vs. shielded input? https://zips.z.cash/zip-0244#s-2d-txin-sig-digest

Reply:
The type of the inputs to this function now ensure that it can only be used in the context of a transparent input; if it's not a transparent input, you have to just sign the txid hash.